Scala ecosystem SDK doc (#1025)
digital-asset/daml: docs/source/app-dev/bindings-scala/index.rst

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Scala bindings
##############

The Scala bindings is a client implementation of the *Ledger API*

> based on `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_.

Reviewer: Since akka-streams is optional, we might not say that it is "based on" it here.

Introduction
============

The Scala bindings library lets you write applications that connect to the Digital Asset distributed ledger using the Scala programming language.

There are two main components:

- Akka Streams-based API

>     The API that you use to send commands to the ledger and receive transactions back.

Reviewer: Likewise, we probably want to mention a third component, the akka-less API.

Author: @stephencompall-DA what is that 3rd component? We have not tested Scala codegen with anything else but akka-based API.

Author: you technically can send generated commands using any existing API, doesn't have to be scala akka-based bindings, can be java rx-based api too. But the 3rd component... is something I am not aware of.

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Scala bindings
##############

> This tutorial provides a basic Scala programmer's introduction to working with Digital Asset distributed ledger,

Reviewer 2: I wouldn't call this a tutorial - a tutorial walks you through step by step how to achieve something. I would just say "this page".

using the Scala programming language and the *Ledger API*.

Introduction
============

The Scala bindings is a client implementation of the *Ledger API*. The Scala bindings library lets you write applications that connect to the Digital Asset distributed ledger using the Scala programming language.

There are two main components:

- Scala codegen
    DAML to Scala code generator. Use this to generate Scala classes from DAML models. The generated Scala code provides a type safe way of creating contracts (:ref:`com.digitalasset.ledger.api.v1.createcommand`) and exercising contract choices (:ref:`com.digitalasset.ledger.api.v1.exercisecommand`).

- Akka Streams-based API
    The API that you use to send commands to the ledger and receive transactions back.

In order to use the Scala bindings, you should be familiar with:

- :doc:`DAML language </daml/reference/index>`
- :doc:`Ledger API </app-dev/grpc/index>`
- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
- `Scala programming language <https://www.scala-lang.org>`_
- :ref:`assistant-manual-building-dars`

> Getting started with sbt

Reviewer 2: I'm a bit confused as to what you're doing here. Are you telling people how to do something, or are you showing them what an example build.sbt file looks like? You need to explicitly tell the reader what the purpose of the information you're giving them is.

========================

> Set up the following dependencies in your ``build.sbt``:

Reviewer 2: What build.sbt? Do you expect them to have this file already?

> Set up the following dependencies in your ``build.sbt``:

Reviewer 2: Do you want to tell them to get the quickstart-scala example first?

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:dependencies>
   :end-before: // </doc-ref:dependencies>

We recommend separating generated code and application code into different modules. There are two modules in the example below:

- ``scala-codegen``
    This modules contains all generated Scala classes.
- ``application``
    This is the application code that makes use of the generated Scala classes.

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:modules>
   :end-before: // </doc-ref:modules>

``scala-codegen`` module uses the following function to generate Scala classes from a DAR file.

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:generate-scala>
   :end-before: // </doc-ref:generate-scala>

> Here is the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`_

Reviewer 2: ```suggestion
You can get the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`__ from the ``daml`` repository on GitHub.
```

I'm a bit anxious about this link - if you move or rewrite the example, it'll just break.

Author: we should not be moving it. I thought the RST doc build tool actually checks if links exist

Reviewer 2: It checks if links internal to the docs exist. It doesn't have a way of checking links to external websites!

Author: I will see if I can change it to point to the symlink inside the docs that I created

Reviewer 2: I don't think that'll work - as far as I'm aware, that code file won't be published, so you won't be able to link to it within the docs site.

You might be able to make it a download link instead? Search "sphinx download directive" for how to do that - you'll want to preview it locally and make sure it works, though

Author: In this case I will leave it for later as a tech debt (issue will be created)... I have a hard deadline to finish the docs by Thursday. We will not be moving the examples anywhere any time soon.

Author: #1134

> Generating Scala Code from Command Line

Reviewer 2: Not sure why the capital letters here?

```suggestion
Generating Scala code from the command line
```

=======================================

The above example demonstrates how to use Scala codegen from **sbt**. You can also call Scala codegen directly

> from a command line. For information on how to build a **DAR** file from a **DAML** model see :ref:`assistant-manual-building-dars`.

Reviewer 2: "For information on how to build a **DAR** file" comes completely out of the blue here - you haven't told anyone why they'd need to care about it at all! 

I'd suggest rewriting this as a set of instructions:

1. If you're not on the latest, upgrade the SDK
2. Download the codegen tool
3. Build the DAR file (this is the right place to link to instructions)
4. Run the codegend tool.

The reason for this is that you need to help people out: you're giving them a bunch of information but it's not in the logical order that they'd need to follow it in, so it makes it much tougher to use this doc.

of the Scala codegen command line interface. Make sure that this versions is aligned with the version of the SDK you used to build the **DAR** file.

To run Scala codegen::

    $ java -jar <parth-to-codegen-main-jar> scala <path-to-DAR-file>=<package-name> \
        --output-directory=<path-to-output-directory> --verbosity=<0|1|2|3|4>

Here is an example, assuming SDK Version: **0.12.17**, DAR file: **./quickstart-scala.dar**,
package name: **com.digitalasset.quickstart.iou.model**, codegen output directory: **./codegen-out** and
verbosity level: **2** (INFO)::

    $ java -jar codegen-main-100.12.17.jar scala ./quickstart-scala.dar=com.digitalasset.quickstart.iou.model \
        --output-directory=./codegen-out --verbosity=2
    ...
    [INFO ] Scala Codegen result:
    Number of generated templates: 3
    Number of not generated templates: 0
    Details:

The output above tells that codegen produced Scala classes for 3 templates without errors (empty ``Details:`` line).

Example code
============

> This section refers to the IOU DAML example from the :doc:`Quickstart guide </getting-started/quickstart>` and

Reviewer 2: You haven't told users what this section is about! Can you write a brief introduction saying what they'll learn from reading it - eg, this section shows you how to use the Scala bindings to achieve common tasks?

`quickstart-scala sbt example <https://github.com/digital-asset/daml/tree/master/language-support/scala/examples/quickstart-scala>`_.

> Please keep in mind that all Scala examples compiled with ``-Xsource:2.13`` **scalac** option.

Reviewer 2: Can you tell me why this is important?

> Create a Contract and Send a Create Command

Reviewer 2: ```suggestion
Create a contract and send a create command
```

Again, no need for capital letters.

Aren't 'creating a contract' and 'sending a create command' the same thing?

-------------------------------------------

To create a Scala class representing an ``Iou`` contract:

> ::

Reviewer 2: You don't need the :: on a separate line, you can have it at the end of the previous one.

>     import com.digitalasset.ledger.client.binding.{Primitive => P}

Reviewer 2: Could you `literalinclude` this from somewhere rather than having it inline? Otherwise there's no guarantee this code won't break.

    import com.digitalasset.quickstart.iou.model.{Iou => M}
    ...
    private val issuer = P.Party("Alice")
    ...

    val iou = M.Iou(
      issuer = issuer,
      owner = issuer,
      currency = "USD",
      amount = BigDecimal("1000.00"),
      observers = List())

To send a corresponding create command:
::

>     val createCmd = iou.create

Reviewer 2: Same comment as above, can you use a literalinclude?

    clientUtil.submitCommand(issuer, issuerWorkflowId, createCmd)

> For more details, please refer to the implementation of `com.digitalasset.quickstart.iou.ClientUtil#submitCommand <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/ClientUtil.scala>`_.

Reviewer 2: What exactly will they find out from reading that? ie - what are the more details?

> Receive a Transaction, Exercise a Choice and Send an Exercise Command

Reviewer 2: ```suggestion
Receive a transaction, exercise a choice and send an exercise command
```

Receive a transaction feels like a separate task to me?

---------------------------------------------------------------------

To receive a transaction and decode a create event for ``IouTransfer`` contract:
::

>     private val newOwner = P.Party("Bob")

Reviewer 2: Same comment as above - please use a literalinclude

>     private val newOwner = P.Party("Bob")

Reviewer 2: Could you also write some comments in this code to explain what it's doing?

Author: I don't think that creating a Party instance needs explanations.

Reviewer 2: I just put the comment here as it was the start of the code snippet - I leave it to you to best decide which bits of the code are worthwhile commenting. But in general just giving some descriptive comments is helpful in documentation - more than you would in your own source code

Author: I think the doc reference comments like the logging calls should be more than enough to figure out what the following code snippet supposed to do:
``` 
    // <doc-ref:subscribe-and-decode-iou-transfer>
    _ <- clientUtil.subscribe(newOwner, offset0, None) { tx =>
      logger.info(s"$newOwner received transaction: $tx")
```
Sorry, I will not add more comments explaining what the example does. I prefer to receive a feed back from the developers looking at the example and make it cleaner if needed.

    ...
    clientUtil.subscribe(newOwner, offset0, None) { tx =>
      logger.info(s"$newOwner received transaction: $tx")
      decodeCreated[M.IouTransfer](tx).foreach { contract: Contract[M.IouTransfer] =>
        logger.info(s"$newOwner received contract: $contract")
        ...
      }
    }

To exercise ``IouTransfer_Accept`` choice on the received ``IourTransfer`` contract and send a corresponding exercise command:
::

>     val exerciseCmd = contract.contractId.exerciseIouTransfer_Accept(actor = newOwner)

Reviewer 2: Same comment as above - please use a literalinclude

    clientUtil.submitCommand(newOwner, newOwnerWorkflowId, exerciseCmd)

> Fore more details, please refer to the `com.digitalasset.quickstart.iou.IouMain#newOwnerAcceptsAllTransfers <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala>`_.

Reviewer 2: Same as above - what will they learn from this? Be specific as to why they should follow the link, what it will help them with.

Receive a transaction, exercise a choice and send an ExerciseCommand
---------------------------------------------------------------------

To receive a transaction as a **newOwner** and decode a :ref:`com.digitalasset.ledger.api.v1.createdevent` for ``IouTransfer`` contract, you need the definition of the **newOwner** ``Party``:

.. literalinclude:: ./code-snippets/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala
   :start-after: // <doc-ref:new-owner-definition>
   :end-before: // </doc-ref:new-owner-definition>

and the following code that handles subscription and decoding:

.. literalinclude:: ./code-snippets/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala
   :start-after: // <doc-ref:subscribe-and-decode-iou-transfer>
   :end-before: // </doc-ref:subscribe-and-decode-iou-transfer>

> To exercise ``IouTransfer_Accept`` choice on the ``IourTransfer`` contract that you received and send a corresponding :ref:`com.digitalasset.ledger.api.v1.exercisecommand`:

Reviewer: ```suggestion
To exercise ``IouTransfer_Accept`` choice on the ``IouTransfer`` contract that you received and send a corresponding :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand`:
```

Author: thanks... but the ref link is no caps.

Author: actually the doc build tool takes references in camel notation too and converts them to all small letters. I like camel... more readable. Going to check the rest of the link in the doc.

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Scala bindings
##############

.. toctree::
  :hidden:

The Scala bindings is a client implementation of the *Ledger API*
based on `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_.

Introduction
************

The Scala bindings library lets you write applications that connect to the Digital Asset distributed ledger using Scala programming language.

There are two main components:

- The Akka Streams based API
    The API to send commands to the ledger and receive transactions back.

- Scala codegen
    DAML to Scala code generator. Generates Scala classes from DAML models. The generated Scala code provides a type safe way of creating contracts (create command) and exercising contract choices (exercise command).

The reader must be familiar with:

- :doc:`DAML language </daml/reference/index>`
- :doc:`Ledger API </app-dev/grpc/index>`
- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
- `Scala programming language <https://www.scala-lang.org>`_

> Getting started with SBT

Reviewer 2: I don't understand - is this describing how to set up SBT in general, or for the quickstart example specifically?

Author: This is SBT specific setup., there is nothing quickstart specific in here.

Author: I am just referring to something that I know works.

> Getting started with SBT

Author: you can use it from maven too, it is just we don't provide an example how to set it up in maven.

************************

> Setting up dependencies in ``build.sbt``:

Reviewer 2: could you rephrase this as an instruction?

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:dependencies>
   :end-before: // </doc-ref:dependencies>

There are two modules in the example below:

- ``scala-codegen``

>     This modules contains generated Scala ADTs.

Reviewer 2: Above you said "classes" - can you pick one term or the other?

Author: ADTs (algebraic data types) are classes. But not all classes are ADTs. I will use classes. Thanks.

- ``application``

>     This is the application code that makes use of the generated Scala ADTs.

Reviewer 2: Is this just needed to get your quickstart example working, or will everyone need this?

Author: yep, everyone will need it. This is a recommended way of setting up an application that uses scala codegen. Two modules, one module that contains generated classes and another module for the application code.

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:modules>
   :end-before: // </doc-ref:modules>

``scala-codegen`` module uses the following function to generate Scala ADTs from a DAR file.

.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
   :start-after: // <doc-ref:generate-scala>
   :end-before: // </doc-ref:generate-scala>

Here is the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`_

Example code
************

> TODO: Explain ADTs here

Reviewer 2: This is missing?

Author: yep, it is still WIP... it is a darf PR :)
